### PR TITLE
feat: Exchange migration

### PR DESCRIPTION
**Pull request by RabbitDoge**
MIGRATION !

### CODE REVIEW

**Comment by matbrgz** [comment]
Is this branch open to contribution? Is there something that I could help?

**Comment by RabbitDoge** [comment]
> Is this branch open to contribution? Is there something that I could help?

The work here is almost finished and is being tested. Better to wait until this branch is merged if you want to contribute.

File: src/config/constants/multicall/abi.json
Comment (ChefHutch) [inline src/config/constants/multicall/abi.json:1]: Do we now have two multicall implementations? https://github.com/pancakeswap/pancake-frontend/blob/2ac9618a63dc6fbb46911c1d4ae9d6318cd0051d/src/hooks/useContract.ts#L210 which uses `src/config/constants/multicall/abi.json`, and `0x1Ee38d535d541c55C9dae27B12edf090C608E6Fb` / `0x301907b5835a2d723Fe3e9E8C5Bc5375d5c1236A`, 

And then also:
 
https://github.com/pancakeswap/pancake-frontend/blob/7b7c12cc84a98ebfba2b0e3b43502b2a712c3033/src/utils/contractHelpers.ts#L129 (used by `src/utils/multicall.ts`) which uses `src/config/abi/Multicall.json` & `0xfF6FD90A470Aaa0c1B8A54681746b07AcdFedc9B` / `0x8F3273Fb89B075b1645095ABaC6ed17B2d4Bc576`

Presuming we need both - it would be good to distinguish this exchange-specific ABI/contract?

File: src/hooks/useBUSDPrice.ts
Comment (ChefHutch) [inline src/hooks/useBUSDPrice.ts:21]: Is there a plan to refactor the continued references to tokens `WETH` / `ETHER` coming from the SDK to reflect BNB/WBNB, and then continue this into the FE so that variables like `ethPairState `, `showETH` etc also refer to the correct ticker?

File: src/state/application/hooks.ts
Comment (ChefHutch) [inline src/state/application/hooks.ts:5]: Could our existing 'current block' implementation could be used instead? https://github.com/pancakeswap/pancake-frontend/blob/7d10afe553d5c8ef6a3b1351c176de397607c249/src/state/hooks.ts#L356.
 
(dispatched by)

https://github.com/pancakeswap/pancake-frontend/blob/7d10afe553d5c8ef6a3b1351c176de397607c249/src/state/hooks.ts#L85

If not, do we refactor other bits of the app to use this block value?

As the `state/application` folder seems to entirely be storing current block state, maybe it should be renamed?

File: src/utils/index.ts
Comment (ChefHutch) [inline src/utils/index.ts:85]: Effectively a duplicate of https://github.com/pancakeswap/pancake-frontend/blob/7b7c12cc84a98ebfba2b0e3b43502b2a712c3033/src/utils/contractHelpers.ts#L56?

File: src/utils/index.ts
Comment (ChefHutch) [inline src/utils/index.ts:46]: Effectively a duplicate of 

https://github.com/pancakeswap/pancake-frontend/blob/66a00e5af08e7203e375a1a0b81173e98ba1fd68/src/utils/truncateWalletAddress.ts#L5

?

File: src/utils/index.ts
Comment (ChefHutch) [inline src/utils/index.ts:12]: If this file `src/utils/index.ts` is used entirely by exchange code, perhaps its name should reflect that?

**Review by ChefHutch — COMMENTED** [review]
Highlighted a few bits in comments - broadly my feedback would be:

- The preview works really well, nice work
- Is there a plan to refactor duplicate code that's now being brought in? Are instances of that duplication already noted down? In some instances it's just a small function (like `getContract`), in others it's larger pieces of state (like `src/state/application` / `state.application.blockNumber`). Maybe we need to maintain multiple versions of some of this logic - but we should distinguish between what is exchange-specific (like in the new `multicall` code, ABI & contract that's being brought in).
- There's also some nice logic coming in that we could make use of in the FE, off the top of my head stuff like `useInterval.ts`, `useBUSDPrice.ts`, `useIsWindowVisible.ts` `useTokenAllowance.ts`. Do we have a plan to migrate the FE over to these/ even just a list of these functions that  we could make use of elsewhere in the FE?

**Review by hachiojidev — APPROVED** [review]

**Comment by RabbitDoge** [comment]
Thanks @ChefHutch  for the feedbacks. Indeed there is a lot of duplicated code now. Refactor would delay the release by a few days or weeks, so it will be part of a continuous improvement process. 
I have tried to remove some duplicated code but so far, here is what's left:
- Duplicated multicall
- Duplicated state block
- Duplicated tokens in constants
- Duplicated hooks (mostly allowance and balances)
- Duplicated contract helpers

**Comment by ChefHutch** [comment]
@RabbitDoge yup, makes total sense nice one 👍

**Pull request merged**
